quinn/src/connection.rs

```diff
@@ -597,15 +600,72 @@ impl Write for BiStream {
             Async::NotReady => Ok(Async::NotReady),
         }
     }
+}
+
+impl io::Write for SendStream {
+    fn write(&mut self, buf: &[u8]) -> io::Result<usize> {
+        match self.poll_write(buf) {
+            Ok(Async::Ready(n)) => Ok(n),
+            Ok(Async::NotReady) => Err(io::Error::new(io::ErrorKind::WouldBlock, "stream blocked")),
+            Err(WriteError::Stopped { error_code }) => Err(io::Error::new(
+                io::ErrorKind::ConnectionReset,
+                format!("stream stopped by peer: error {}", error_code),
+            )),
+            Err(WriteError::ConnectionClosed(e)) => Err(io::Error::new(
+                io::ErrorKind::ConnectionAborted,
+                format!("connection closed: {}", e),
+            )),
+            Err(WriteError::UnknownStream) => {
+                Err(io::Error::new(io::ErrorKind::Other, "unknown stream"))
+            }
+        }
+    }
 
-    fn reset(&mut self, error_code: u16) {
+    fn flush(&mut self) -> io::Result<()> {
+        Ok(())
+    }
+}
+
+impl AsyncWrite for SendStream {
+    fn shutdown(&mut self) -> Poll<(), io::Error> {
+        self.poll_finish().map_err(|e| {
+            io::Error::new(
+                io::ErrorKind::ConnectionAborted,
+                format!("connection closed: {}", e),
+            )
+        })
+    }
+}
+
+impl Drop for SendStream {
+    fn drop(&mut self) {
         let mut conn = self.conn.lock().unwrap();
-        conn.inner.reset(self.stream, error_code);
+        if conn.error.is_some() {
+            return;
+        }
+        if !self.finished {
+            conn.inner.reset(self.stream, 0);
+        }
         conn.notify();
     }
 }
 
-impl Read for BiStream {
+/// A stream that can only be used to receive data
+pub struct RecvStream {
+    conn: ConnectionRef,
+    stream: StreamId,
+    recvd: bool,
+}
+
+impl RecvStream {
+    fn new(conn: ConnectionRef, stream: StreamId) -> Self {
+        Self {
+            conn,
+            stream,
+            recvd: false,
+        }
+    }
+
     fn poll_read_unordered(&mut self) -> Poll<(Bytes, u64), ReadError> {
```

Comment:
This and `poll_read` should be `pub`, and be documented by the comments that were on the removed `trait Read`.

Reply:
Fixed.

Reply:
Looks like it's still missing `pub`.